src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -105,5 +107,37 @@ private void AppendPropertyName(StringBuilder sb, string propertyName)
                 }
             }
         }
+
+        internal ResolvedReferenceHandling PreserveReference(object value, out string referenceId)
+        {
+            // Avoid emitting metadata to value types.
+            Type currentType = Current.JsonPropertyInfo?.DeclaredPropertyType ?? Current.JsonClassInfo.Type;
+            if (currentType.IsValueType)
+            {
+                referenceId = default;
+                return ResolvedReferenceHandling.None;
+            }
+
+            if (GetPreservedReference(value, out referenceId))
+            {
+                return ResolvedReferenceHandling.IsReference;
+            }
+
+            return ResolvedReferenceHandling.Preserve;
+        }
+
+        // true if reference already exists; otherwise, false;
+        public bool GetPreservedReference(object value, out string id)
+        {
+            if (_referenceResolver == null)
+            {
+                _referenceResolver = new DefaultReferenceResolver();
+            }
+
+            bool handling = _referenceResolver.IsReferenced(value);
+            id = _referenceResolver.GetReference(value);
+
+            return handling;
```

Comment:
Just as a comment, I was trying to stick as much as possible to https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Serialization_IReferenceResolver.htm in case we ever expose it.